File: python/sorting/quick_sort.py
def swap(arr, a, b):
    arr[a], arr[b] = arr[b], arr[a]
    
def partition(arr, si, ei):
    i= si-1 
    pivot= arr[ei] 

    for j in range(si, ei):
        if arr[j]<= pivot:
            i+=1
            arr[j], arr[i] = arr[i], arr[j]
            
    i+=1 
    print(pivot,arr, ei, i)
    arr[ei], arr[i] = arr[i], arr[ei]
    return i

def quickSort(arr, si, ei):
    if si < ei:
        pi = partition(arr, si, ei)
        quickSort(arr, si, pi-1)
        quickSort(arr, pi+1, ei)



arr = [7,2,9,1,5,3]

quickSort(arr, 0, len(arr)-1 )
print(arr)

File: java/Sorting/QuickSort.java
import java.util.*;

public class QuickSort{
    private static void quickSort(int[] a, int si, int ei){
        if(si==ei){
            return;
        }
        int pi = partition(a, si, ei);
        quickSort(a, si, pi-1);
        quickSort(a, pi+1, ei);
    }

    private static int partition(int[] a, int si, ei){
        int pivot=a[ei], i=-1;

        for(int j=si; j<ei; j++){
            if(a[j]< pivot){
                i++;
                int temp = a[i];
                a[i] = a[j];
                a[j] = temp;
            }
        }
        i++;
        temp = a[i];
        a[i] = pivot;
        a[ei] = temp;
        return i;
    }
    public static void main(String args[]){
        Scanner sc= new Scanner(System.in);

        System.out.println("Enter size of array");
        int n= sc.nextInt();

        System.out.println("Enter Array Elements");
        int[] unsortedArray = new int[n];
        for(int i=0; i<n; i++){
            unsortedArray[i] = sc.nextInt();
        }

        sc.close();

        quickSort(unsortedArray, 0, n-1);
        
        for(int i=0; i<n; i++){
            System.out.print(unsortedArray[i]+ " ");
        }
    }
}